s3.go

```diff
@@ -1,212 +1,352 @@
-
-	s.buck.Sign(req)
-	resp, err := s.buck.Do(req)
+func (s *S3Bucket) GetSize(k ds.Key) (size int, err error) {
+	resp, err := s.S3.HeadObject(&s3.HeadObjectInput{
+		Bucket: aws.String(s.Bucket),
+		Key:    aws.String(s.s3Path(k.String())),
+	})
 	if err != nil {
-		return false, err
+		if s3Err, ok := err.(awserr.Error); ok && s3Err.Code() == "NotFound" {
+			return -1, ds.ErrNotFound
+		}
+		return -1, err
 	}
-	defer resp.Body.Close()
-	io.Copy(ioutil.Discard, resp.Body)
+	return int(*resp.ContentLength), nil
+}
 
-	switch resp.StatusCode {
-	case 200, 304:
-		return true, nil
-	case 404:
-		return false, nil
-	default:
-		return false, fmt.Errorf(resp.Status)
-	}
+func (s *S3Bucket) Delete(k ds.Key) error {
+	_, err := s.S3.DeleteObject(&s3.DeleteObjectInput{
+		Bucket: aws.String(s.Bucket),
+		Key:    aws.String(s.s3Path(k.String())),
+	})
+	return parseError(err)
 }
 
 func (s *S3Bucket) Query(q dsq.Query) (dsq.Results, error) {
-	if !q.KeysOnly {
-		return nil, fmt.Errorf("s3 datastore doesnt support returning values in query, only keys")
-	}
-
-	if q.Prefix != "" {
-		return nil, fmt.Errorf("s3ds doesnt support prefixes")
-	}
-
 	if q.Orders != nil || q.Filters != nil {
-		return nil, fmt.Errorf("s3ds doesnt support filters or orders")
+		return nil, fmt.Errorf("s3ds: filters or orders are not supported")
 	}
 
-	req, err := http.NewRequest("GET", "https://"+s.buck.Name+"."+s.buck.Domain, nil)
-	if err != nil {
-		return nil, err
+	limit := q.Limit + q.Offset
+	if limit == 0 || limit > listMax {
+		limit = listMax
 	}
 
-	s.buck.Sign(req)
-
-	resp, err := s.buck.Do(req)
+	resp, err := s.S3.ListObjectsV2(&s3.ListObjectsV2Input{
+		Bucket:  aws.String(s.Bucket),
+		Prefix:  aws.String(s.s3Path(q.Prefix)),
+		MaxKeys: aws.Int64(int64(limit)),
+	})
 	if err != nil {
 		return nil, err
 	}
 
-	dec := xml.NewDecoder(resp.Body)
-	var nextiskey bool
-
+	index := q.Offset
 	nextValue := func() (dsq.Result, bool) {
-		for {
-			tok, err := dec.Token()
+		for index >= len(resp.Contents) {
+			if !*resp.IsTruncated {
+				return dsq.Result{}, false
+			}
+
+			resp, err = s.S3.ListObjectsV2(&s3.ListObjectsV2Input{
+				Bucket:            aws.String(s.Bucket),
+				Prefix:            aws.String(s.s3Path(q.Prefix)),
+				Delimiter:         aws.String("/"),
+				MaxKeys:           aws.Int64(listMax),
+				ContinuationToken: resp.NextContinuationToken,
+			})
 			if err != nil {
-				if err == io.EOF {
-					return dsq.Result{}, false
-				}
 				return dsq.Result{Error: err}, false
 			}
 
-			switch tok := tok.(type) {
-			case xml.StartElement:
-				if tok.Name.Local == "Key" {
-					nextiskey = true
-				}
-			case xml.CharData:
-				if nextiskey {
-					kval := string(tok)
-					nextiskey = false
-					dec.Skip()
-
-					if strings.HasPrefix(kval, ".md5/") {
-						continue
-					}
-					res := dsq.Result{Entry: dsq.Entry{Key: kval}}
-					return res, true
-				}
+			index -= len(resp.Contents)
```

Comment:
Actually, I think this logic is wrong. Shouldn't this go *above* the `ListObjectsV2` call? That is, if `index >= len(resp.Contents)`, we want to *skip* those contents.

Reply:
You're right, not sure why I had it down there.